Add Multi Instance Support (#3)
projectsyn/component-vault: docs/modules/ROOT/pages/tutorials/install.adoc

= Install

> To install vault using this component, there are a few things to decide and prepare.

Reviewer: ```suggestion
To install Vault using this component, there are a few things to decide and prepare.
```

== Backups

> By default backups using https://k8up.io/k8up/1.0.0/index.html[k8up] are enabled.

Reviewer: ```suggestion
By default, backups using https://k8up.io/k8up/1.0.0/index.html[k8up] are enabled.
```

For this to work `k8up` needs to be installed on the same cluster using the https://github.com/projectsyn/component-backup-k8up[k8up component].

> By default the component uses the following configuration:

Reviewer: ```suggestion
By default, the component uses the following configuration:
```

[source,yaml]
----
backup:
  enabled: true
  schedule: '*/13 * * * *'
  keepjobs: 5
  password: '?{vaultkv:${customer:name}/${cluster:name}/vault/${_instance}/backup/password}'
  bucket:
    name: '${_instance}-backup'
    accesskey: '?{vaultkv:${customer:name}/${cluster:name}/vault/${_instance}/backup/s3_access_key}'
    secretkey: '?{vaultkv:${customer:name}/${cluster:name}/vault/${_instance}/backup/s3_secret_key}'
----

> This means it expects that there is a bucket with the name `+${INSTANCE}-backup+` and that three secrets need to be added to the deploying Vault.

Reviewer: ```suggestion
The backup configuration expects that there is a bucket with the name `+${INSTANCE}-backup+`.

The three referenced secrets need to be added to the Vault instance associated with the Lieutenant instance managing the cluster on which this component is installed:
```

> * The password to use for backup encryption to the key `+clusters/kv/${TENANT_ID}/${CLUSTER_ID}/vault/${INSTANCE}/backup/password}+`
> * The access key for the bucket to to the key `+clusters/kv/${TENANT_ID}/${CLUSTER_ID}/vault/${INSTANCE}/backup/s3_access_key}+`
> * The secret key for the bucket to to the key `+clusters/kv/${TENANT_ID}/${CLUSTER_ID}/vault/${INSTANCE}/backup/s3_secret_key}+`

Reviewer: ```suggestion
* The password to use for backup encryption in key `+clusters/kv/${TENANT_ID}/${CLUSTER_ID}/vault/${INSTANCE}/backup/password}+`
* The access key for the bucket in key `+clusters/kv/${TENANT_ID}/${CLUSTER_ID}/vault/${INSTANCE}/backup/s3_access_key}+`
* The secret key for the bucket in key `+clusters/kv/${TENANT_ID}/${CLUSTER_ID}/vault/${INSTANCE}/backup/s3_secret_key}+`
```

> If you only deploy a single vault to the cluster, `+${INSTANCE}+` will simply be `+vault+`.

Reviewer: ```suggestion
Unless you explicitly configure multiple instances of the component, a single vault instance will be deployed.
In this case, `+${INSTANCE}+` will simply be `+vault+`.
```

> See <<Multiple Instances>> for more details on this parameter.

Reviewer: ```suggestion
See <<Multiple Instances>> for more details on how to deploy multiple Vault instances using this component.
```

== LDAP

By default, only the backup service account and the root user has access to vault.
You have the option to add more authentication methods and policies. 

> Here is an example to add vault access to all people in an LDAP group.

Reviewer: ```suggestion
Here is an example to add Vault access to all people in an LDAP group.
```

[source,yaml]
----
config:
  auth:
    - type: ldap
      description: LDAP auth
      options:
        listing_visibility: "unauth"
      config:
        url: ldaps://ldap.server.net:636
        binddn: "uid=vault-service,ou=Special Users,dc=server,dc=net"
        bindpass: '?{vaultkv:${customer:name}/${cluster:name}/vault/${_instance}/ldappass}'
        userattr: uid
        userdn: "ou=vault,ou=Service Access,ou=Views,dc=server,dc=net"
        groupdn: "ou=Groups,dc=vshn,dc=net"
        groupattr: cn
      groups:
        vault-root:
          policies: cluster-root
  policies:
    - name: cluster-root
      rules: |
        path "clusters/*" {
          capabilities = ["create", "read", "update", "delete", "list"]
        }
----

Under the hood, this component uses `+bank-vaults+` to configure vault.
https://banzaicloud.com/docs/bank-vaults/external-configuration/[This reference] provides more details on how to configure policies and authentication methods.

== Lieutenant Integration

> If vault interacts with a Lieutenant instance, we need to add some policies and authentication methods.

Reviewer: ```suggestion
If your Vault should interact with a Lieutenant instance, we need to add some policies and authentication methods.
```

Maybe worth to reference `component-lieutenant` here?

[source,yaml]
----
config:
  policies:
    - name: syn-cluster
      rules: | # Allow to get cluster's own secrets

>         path "clusters/kv/data/+/{{identity.entity.aliases.auth_kubernetes_5b636428.metadata.service_account_name}}/*" {

Reviewer: ```suggestion
        path "clusters/kv/data/+/{{identity.entity.aliases.auth_kubernetes_5b636428.metadata.service_account_name}}/*" {
```

Probably need to make a note that the `auth_kubernetes_<id>` part of this path needs to be identified after activating Kubernetes auth in Vault.

Also note that this config expects that Vault and Lieutenant are deployed in the same cluster.

          capabilities = ["read"]
        }
    - name: lieutenant-operator
      rules: |
        path "clusters/kv/data/*" {
          capabilities = ["read", "create", "update", "delete"]
        }
        path "clusters/kv/metadata/*" {
          capabilities = ["read", "create", "update", "delete", "list"]
        }
        path "clusters/kv/delete/*" {
          capabilities = ["update"]
        }
  auth:
    - type: kubernetes
      roles:
        - name: syn-cluster
          bound_service_account_names: "*"
          bound_service_account_namespaces: lieutenant-prod
          policies: syn-cluster
          ttl: 1h
        - name: lieutenant-operator
          bound_service_account_names: lieutenant-operator
          bound_service_account_namespaces: lieutenant-prod
          policies: lieutenant-operator
          ttl: 1h
----

> This allows the `+lieutenant-operator+` to read and write secrets.

Reviewer: ```suggestion
The `syn-cluster` policy allows each cluster to reveal its own secrets.
The `lieutenant-operator` policy the `+lieutenant-operator+` to read and write secrets.
```

If service accounts or namespaces differ in your deployment, you will need to update the snippet above

== Multiple Instances

> This component is able to deploy multiple instances of vault on the same cluster.

Reviewer: ```suggestion
This component is able to deploy multiple instances of Vault on the same cluster.
```

> You can deploy multiple instances of vault the following way

Reviewer: ```suggestion
You can deploy multiple instances of Vault by instantiating the component.
The following configuration creates two Vault instances named `vault-prod` and `vault-test`.
```

[source,yaml]
----
applications:
  - vault as vault-prod
  - vault as vault-test
parameters:
  vault_prod:
    config:
      auth:
        - type: ldap
          description: LDAP auth
          options:
            listing_visibility: "unauth"
          config:
            url: ldaps://ldap.server.net:636
            binddn: "uid=vault-service,ou=Special Users,dc=server,dc=net"
            bindpass: '?{vaultkv:${customer:name}/${cluster:name}/vault/${_instance}/ldappass}'
            userattr: uid
            userdn: "ou=vault,ou=Service Access,ou=Views,dc=server,dc=net"
            groupdn: "ou=Groups,dc=vshn,dc=net"
            groupattr: cn
          groups:
            vault-root:
              policies: cluster-root
      policies:
        - name: cluster-root
          rules: |
            path "clusters/*" {
              capabilities = ["create", "read", "update", "delete", "list"]
            }
  vault_test:
    backup:
      enabled: false
----

This will deploy a production vault with LDAP access named `vault-prod` to the namespace `vault-prod`.
And a test vault without backups named `vault-test` to the namespace `vault-test`.

> This is where it `+${INSTANCE}+` parameter is defined.

Reviewer: ```suggestion
When instantiating a component the `+${INSTANCE}+` parameter is set to the instance's name.
```

> For the production vault the parameter is set to `vault-prod` and for the test vault to `vault-test`.

Reviewer: ```suggestion
For the production Vault, the parameter is set to `vault-prod` and for the test Vault it is set to `vault-test`.
```

> There are some things to consider when deploying multiple instances of vault:

Reviewer: ```suggestion
There are some things to consider when deploying multiple instances of Vault:
```

* No two instances are allowed to have the same name.
This includes instances of other components.
You should never name an instance the same name as other components.

> Naming your vault instance `argocd` can break in unexpected ways. 

Reviewer: ```suggestion
Naming your Vault instance `argocd` can break in unexpected ways. 
```
